pkg/network/protocols/kafka/stats.go

```diff
@@ -54,8 +61,26 @@ func NewRequestStats() *RequestStats {
 
 // RequestStat stores stats for Kafka requests to a particular key
 type RequestStat struct {
-	Count      int
-	StaticTags uint64
+	// this field order is intentional to help the GC pointer tracking
+	Latencies *ddsketch.DDSketch
+	// Note: every time we add a latency value to the DDSketch, it's possible for the sketch to discard that value
+	// (ie if it is outside the range that is tracked by the sketch). For that reason, in order to keep an accurate count
+	// the number of http transactions processed, we have our own count field (rather than relying on DDSketch.GetCount())
+	Count int
+	// This field holds the value (in nanoseconds) of the first HTTP request
+	// in this bucket. We do this as optimization to avoid creating sketches with
+	// a single value. This is quite common in the context of HTTP requests without
+	// keep-alives where a short-lived TCP connection is used for a single request.
```

Comment:
This is taking about HTTP, does this rationale hold for Kafka too?

Reply:
That's a good observation. From what I've seen, the answer is no; most modern Kafka clients are reusing sockets and setting keepalives. An example of this can be found in the [Apache Kafka Java Client](https://github.com/apache/kafka/blob/901c656a1da16c8473200c93e57991004fc1371d/clients/src/main/java/org/apache/kafka/common/network/Selector.java#L303)

Reply:
I'll work on removing the `FirstLatencySample` to simplify the code, as it appears to be an unnecessary optimization

Reply:
After discussing with @guyarb, to fully understand the performance impact, I plan to:
1. Determine the difference between creating a ddsketch and a simple FirstLatency setup.
2. Estimate how often the FirstLatency scenario occurs for a connection in the Datadog environment. To do this, we'll need to add a metric for the number of Kafka requests and compare it to the number of Kafka connections.

Reply:
Implemented the following benchmark tests:

```go
func benchmarkCreateRequestStat(b *testing.B, useLatencies bool) {
	b.ReportAllocs()
	b.ResetTimer()

	latencySample := 3.2
	count := 10

	for i := 0; i < b.N; i++ {
		if useLatencies {
			latencies, err := ddsketch.NewDefaultDDSketch(0.01)
			require.NoError(b, err)
			require.NoError(b, latencies.AddWithCount(latencySample, float64(count)))

			_ = kafka.RequestStat{
				Latencies: latencies,
				Count:     count,
			}
		} else {
			_ = kafka.RequestStat{
				FirstLatencySample: latencySample, // Example latency value
				Count:              count,
			}
		}
	}
}

func BenchmarkCreateRequestStatWithLatencies(b *testing.B) {
	benchmarkCreateRequestStat(b, true)
}

func BenchmarkCreateRequestStatWithFirstLatencySample(b *testing.B) {
	benchmarkCreateRequestStat(b, false)
}

Reply:
The benchmark results are as follows:
`BenchmarkCreateRequestStatWithLatencies   	 1811176	       664.7 ns/op	     784 B/op	       8 allocs/op`
`BenchmarkCreateRequestStatWithFirstLatencySample   	1000000000	         0.3180 ns/op	       0 B/op	       0 allocs/op`

Reply:
## Results Breakdown

**BenchmarkCreateRequestStatWithLatencies:**

- **Operations per second (`ns/op`)**: 664.7 ns/op  
  On average, each operation takes approximately 664.7 nanoseconds.

- **Memory allocated per operation (`B/op`)**: 784 B/op  
  Each operation allocates around 784 bytes of memory.

- **Allocations per operation (`allocs/op`)**: 8 allocs/op  
  Each operation results in 8 memory allocations.

**BenchmarkCreateRequestStatWithFirstLatencySample:**

- **Operations per second (`ns/op`)**: 0.3180 ns/op  
  On average, each operation takes approximately 0.318 nanoseconds. This is extraordinarily fast, likely indicating that the benchmark is measuring a minimal or zero-cost operation.

- **Memory allocated per operation (`B/op`)**: 0 B/op  
  Each operation does not allocate any memory.

- **Allocations per operation (`allocs/op`)**: 0 allocs/op  
  Each operation results in no memory allocations.

## Interpretation

**Performance:**

The `BenchmarkCreateRequestStatWithFirstLatencySample` is significantly faster (0.318 ns/op vs. 664.7 ns/op) compared to `BenchmarkCreateRequestStatWithLatencies`. This is expected because storing a single float value (the `FirstLatencySample`) is much simpler and less costly than managing a `DDSketch` for latencies.

**Memory Usage:**

The `BenchmarkCreateRequestStatWithFirstLatencySample` uses no additional memory (0 B/op and 0 allocs/op), indicating that it’s very efficient in terms of memory usage.

The `BenchmarkCreateRequestStatWithLatencies` allocates 784 bytes per operation and performs 8 allocations, indicating higher memory overhead.

Reply:
We don't currently have an effective way to estimate how often Kafka connections have a single request per connection, which is relevant for the FirstLatencySample optimization. We'll need to add a metric to track how frequently the FirstLatencySample is used and test this in staging and production. Since we lack conclusive data to support either approach at the moment, I'll proceed with the FirstLatencySample optimization for now and we can revisit it in a future task